catalog/core/catalog-core-subscriptionstore/subscriptionstore-impl/src/main/java/org/codice/ddf/catalog/subscriptionstore/SubscriptionContainerImpl.java

```diff
@@ -0,0 +1,396 @@
+     * <ol>
+     * <li>The subscription being identified exists in the cache</li>
+     * <li>The subscription's type matches the identifier's type</li>
+     * <li>The subscription is registered with OSGi</li>
+     * </ol>
+     * <p>
+     * One impact of an incorrect implementation of {@link SubscriptionFactory} will be incorrect
+     * {@code null} results returned for client traffic that occurs during service initialization time.
+     * This is because queries for non-registered subscriptions that do exist but haven't been initialized
+     * yet return null. Subscription container clients that expose external services must ensure their
+     * factory is available prior to, or at the same time as, those exposed services.
+     *
+     * @see SubscriptionFactory
+     */
+    @Nullable
+    @Override
+    public synchronized Subscription get(SubscriptionIdentifier identifier) {
+        validateIdentifier(identifier);
+        String subscriptionId = identifier.getId();
+        String type = identifier.getTypeName();
+
+        CachedSubscription cachedSubscription = subscriptionsCache.get(subscriptionId);
+
+        if (cachedSubscription == null || cachedSubscription.isNotType(type)
+                || cachedSubscription.isNotRegistered()) {
+            return null;
+        }
+
+        return cachedSubscription.getSubscription()
+                .orElseThrow(() -> new SubscriptionStoreException("Could not get subscription. "));
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Register the subscription and write to the cache. If there is a problem, rollback the registration.
+     */
+    @Override
+    public synchronized SubscriptionIdentifier insert(Subscription subscription,
+            MarshalledSubscription marshalledSubscription, SubscriptionType type) {
+        validateSubscription(subscription);
+        validateSerialized(marshalledSubscription);
+        validateType(type);
+
+        SubscriptionMetadata metadata = new SubscriptionMetadata(type.getTypeName(),
+                marshalledSubscription.getFilter(),
+                marshalledSubscription.getCallbackAddress());
+
+        doInsert(subscription, metadata);
+        return metadata;
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Updates are performed by a delete first, then an insertion using the same id. The operations
+     * are identical to {@link #insert(Subscription, MarshalledSubscription, SubscriptionType)}
+     * and to {@link #delete(SubscriptionIdentifier)} with the exception of preserving the id in the
+     * {@link SubscriptionMetadata}.
+     */
+    @Override
+    public synchronized void update(Subscription subscription,
+            MarshalledSubscription marshalledSubscription, SubscriptionIdentifier identifier) {
+        validateSubscription(subscription);
+        validateSerialized(marshalledSubscription);
+        validateContainment(identifier, "update");
+
+        String subscriptionId = identifier.getId();
```

Comment:
I think it actually might work. Since I'm already type-checking the subscriptions. This would simplify the logic since the subscription type is actually helping us uniquely identify them.